geoopt/manifolds/symmetric_positive_definite.py

```diff
@@ -0,0 +1,385 @@
+
+    References
+    ----------
+    - https://github.com/pymanopt/pymanopt/blob/master/pymanopt/manifolds/psd.py
+    - https://github.com/dalab/matrix-manifolds/blob/master/graphembed/graphembed/manifolds/spd.py
+
+    Parameters
+    ----------
+    ndim : int
+        number of trailing dimensions treated as matrix dimensions.
+        All the operations acting on such as inner products, etc
+        will respect the :attr:`ndim`.
+    """
+
+    __scaling__ = Manifold.__scaling__.copy()
+    name = "SymmetricPositiveDefinite"
+    ndim = 0
+    reversible = True
+    defaulf_metric: str = "AIM"
+
+    def __init__(self, ndim=2, defaulf_metric: str = defaulf_metric):
+        super().__init__()
+        self.ndim = ndim
+        self.defaulf_metric = defaulf_metric
+
+    def _funcm(
+        self, x: torch.Tensor, func: Callable[[torch.Tensor], torch.Tensor]
+    ) -> torch.Tensor:
+        """Apply function to symmetric matrix.
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            symmetric matrix
+        func : Callable[[torch.Tensor], torch.Tensor]
+            function to apply
+
+        Returns
+        -------
+        torch.Tensor
+            symmetric matrix with function applied to
+        """
+        e, v = torch.symeig(x, eigenvectors=True)
+        return v @ torch.diag_embed(func(e)) @ v.transpose(-1, -2)
+
+    def _expm(self, x: torch.Tensor, using_native=False) -> torch.Tensor:
+        """Symmetric matrix exponent.
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            symmetric matrix
+        using_native : bool, optional
+            if using native matrix exponent `torch.matrix_exp`, by default False
+
+        Returns
+        -------
+        torch.Tensor
+            exp(x)
+        """
+        if using_native:
+            return torch.matrix_exp(x)
+        else:
+            return self._funcm(x, torch.exp)
+
+    def _logm(self, x: torch.Tensor) -> torch.Tensor:
+        """Symmetric matrix logarithm.
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            symmetric matrix
+
+        Returns
+        -------
+        torch.Tensor
+            log(x)
+        """
+        return self._funcm(x, torch.log)
+
+    def _sqrtm(self, x: torch.Tensor) -> torch.Tensor:
+        """Symmetric matrix square root .
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            symmetric matrix
+
+        Returns
+        -------
+        torch.Tensor
+            sqrt(x)
+        """
+        return self._funcm(x, torch.sqrt)
+
+    def _invm(self, x: torch.Tensor) -> torch.Tensor:
+        """Symmetric matrix inverse.
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            symmetric matrix
+
+        Returns
+        -------
+        torch.Tensor
+            x^{-1}
+        """
+        return self._funcm(x, torch.reciprocal)
+
+    def _inv_sqrtm1(self, x: torch.Tensor) -> torch.Tensor:
```

Comment:
Well, I believe that is faster since it fuses 2 ops. Am I correct?

Reply:
Also, the best place to put this is batch_linalg file

Reply:
> Well, I believe that is faster since it fuses 2 ops. Am I correct?

Yeah, since some operations need both inverse and inverse square root, it's better to fuse these two operations since they both require eigendecomposition.